course4/homework-unique-pointer/refcounted.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include<iostream>
+
+namespace course{
+
+template <typename T>
+class RefcountingPointer
+{
+private:
+    T* payload;
+    uint* count;
+    RefcountingPointer() = default;
+    void decrement()
+    {
+        if(count == nullptr)
+        {
+            return;
+        }
+
+        if(--(*count) == 0)
+        {
+            std::cout << "Deleting: " << (void*)payload << std::endl;
```

Comment:
These are good for debugging during development, but you should remove them when you are done.
You should be able to wrtie automatic unit-test.